### ISSUE TITLE
External: iOS 13+ will not play audio when silent mode on

### ISSUE THREAD

**Issue by Wackytoaster** [Scirra/Construct-3-bugs#3477]
EDIT: Related webkit bug https://bugs.webkit.org/show_bug.cgi?id=204261

iOS 13+ will not play audio at all.
https://forums.developer.apple.com/message/393961#393961

I´m reporting it here because maybe there is a simple workaround other than "hoping apple will maybe fix it in the future"

> Unfortunately, from iOS 13 this ability is under sytem entitlements, com.apple.multitasking.systemappassertions and com.apple.multitasking.unlimitedassertions, you can try it in simulator, no error in log. **There is no way to launch it on device thought, as you need an Apple issued provision profile. So, all we can do is to hope that some future release will fix this.**

## Attach a .c3p

[audioBroken.c3p.zip](https://github.com/Scirra/Construct-3-bugs/files/3842779/audioBroken.c3p.zip)

## Steps to reproduce

1. Build in xcode
2. Run on device with ios 13+
3. Tap square to play audio

## Observed result

No audio

## Expected result

Audio

## More details

<details><summary>debug log</summary>

2019-11-13 18:51:21.640624+0100 New project[1471:271574] [Process] kill() returned unexpected error 1\
2019-11-13 18:51:21.643604+0100 New project[1471:271574] [Process] kill() returned unexpected error 1\
2019-11-13 18:51:21.647210+0100 New project[1471:271866] [assertion] Error acquiring assertion: <NSError: 0x281c01c20; domain: RBSAssertionErrorDomain; code: 2; reason: "Client is missing required entitlement"> \{\
    userInfo = \{\
        RBSAssertionAttribute = <RBSLegacyAttribute: 0x15fde6db0; requestedReason: MediaPlayback; reason: MediaPlayback; flags: PreventTaskSuspend | PreventTaskThrottleDown | WantsForegroundResourcePriority>;\
    \}\
\}\
2019-11-13 18:51:21.647392+0100 New project[1471:271866] [ProcessSuspension]  0x1089fb2a0 - ProcessAssertion() PID 1471 Unable to acquire assertion for process with PID 1471\
2019-11-13 18:51:21.647626+0100 New project[1471:271574] [Process] kill() returned unexpected error 1\
2019-11-13 18:51:21.647753+0100 New project[1471:271574] [Process] kill() returned unexpected error 1\
2019-11-13 18:51:21.647996+0100 New project[1471:271574] [Process] kill() returned unexpected error 1\
2019-11-13 18:51:21.648312+0100 New project[1471:271574] [Process] kill() returned unexpected error 1\
2019-11-13 18:51:21.648692+0100 New project[1471:271574] [Process] kill() returned unexpected error 1\
2019-11-13 18:51:21.649248+0100 New project[1471:271574] [Process] kill() returned unexpected error 1\
2019-11-13 18:51:21.650462+0100 New project[1471:271574] [Process] kill() returned unexpected error 1\
2019-11-13 18:51:21.652894+0100 New project[1471:271866] [assertion] Error acquiring assertion: <NSError: 0x281c05140; domain: RBSAssertionErrorDomain; code: 2; reason: "Client is missing required entitlement"> \{\
    userInfo = \{\
        RBSAssertionAttribute = <RBSLegacyAttribute: 0x15fdbce80; requestedReason: MediaPlayback; reason: MediaPlayback; flags: PreventTaskSuspend | PreventTaskThrottleDown | WantsForegroundResourcePriority>;\
    \}\
\}\
2019-11-13 18:51:21.658110+0100 New project[1471:271574] [Process] kill() returned unexpected error 1\
2019-11-13 18:51:21.659410+0100 New project[1471:271574] [Process] kill() returned unexpected error 1\
2019-11-13 18:51:21.661796+0100 New project[1471:271574] [Process] kill() returned unexpected error 1\
2019-11-13 18:51:21.664235+0100 New project[1471:271859] [assertion] Error acquiring assertion: <NSError: 0x281c057d0; domain: RBSAssertionErrorDomain; code: 2; reason: "Client is missing required entitlement"> \{\
    userInfo = \{\
        RBSAssertionAttribute = <RBSLegacyAttribute: 0x15fde6db0; requestedReason: MediaPlayback; reason: MediaPlayback; flags: PreventTaskSuspend | PreventTaskThrottleDown | WantsForegroundResourcePriority>;\
    \}\
\}\
2019-11-13 18:51:21.664381+0100 New project[1471:271859] [ProcessSuspension]  0x1089fb340 - ProcessAssertion() PID 1471 Unable to acquire assertion for process with PID 1471\
2019-11-13 18:51:21.664603+0100 New project[1471:271574] [ProcessSuspension] 0x1089fb340 - ProcessAssertion::processAssertionWasInvalidated()\
2019-11-13 18:51:21.666864+0100 New project[1471:271574] [Process] kill() returned unexpected error 1\
2019-11-13 18:51:21.667756+0100 New project[1471:271859] [assertion] Error acquiring assertion: <NSError: 0x281c059b0; domain: RBSAssertionErrorDomain; code: 2; reason: "Client is missing required entitlement"> \{\
    userInfo = \{\
        RBSAssertionAttribute = <RBSLegacyAttribute: 0x15fe9ad90; requestedReason: MediaPlayback; reason: MediaPlayback; flags: PreventTaskSuspend | PreventTaskThrottleDown | WantsForegroundResourcePriority>;\
    \}\
\}\
2019-11-13 18:51:21.667861+0100 New project[1471:271859] [ProcessSuspension]  0x1089fb390 - ProcessAssertion() PID 1471 Unable to acquire assertion for process with PID 1474\
2019-11-13 18:51:21.667983+0100 New project[1471:271574] [ProcessSuspension] 0x1089fb390 - ProcessAssertion::processAssertionWasInvalidated()\
2019-11-13 18:51:21.669585+0100 New project[1471:271574] [Process] kill() returned unexpected error 1\
2019-11-13 18:51:21.669740+0100 New project[1471:271574] [Process] kill() returned unexpected error 1\
2019-11-13 18:51:21.669809+0100 New project[1471:271574] [Process] kill() returned unexpected error 1\

</details>

**Affected browsers/platforms:** 

**First affected release:** 

## System details

<details><summary>View details</summary>

Platform information
Browser: Chrome
Browser version: 78.0.3904.97
Browser engine: Chromium
Browser architecture: 64-bit
Context: webapp
Operating system: Windows
Operating system version: 10
Operating system architecture: 64-bit
Device type: desktop
Device pixel ratio: 1
Logical CPU cores: 8
Approx. device memory: 8 GB
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.97 Safari/537.36
C3 release: r175 (beta)
Language setting: en-US

Local storage
Storage quota (approx): 29 gb
Storage usage (approx): 1.3 gb (4.6%)
Persistant storage: Yes

Browser support notes
This list contains missing features that are not required, but could improve performance or user experience if supported.

Nothing is missing. Everything is OK!
WebGL information
Version string: WebGL 2.0 (OpenGL ES 3.0 Chromium)
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: yes
Supports highp precision: yes
Vendor: Google Inc.
Renderer: ANGLE (NVIDIA GeForce GTX 1070 Direct3D11 vs_5_0 ps_5_0)
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 1024
Extensions:

EXT_color_buffer_float
EXT_disjoint_timer_query_webgl2
EXT_float_blend
EXT_texture_filter_anisotropic
KHR_parallel_shader_compile
OES_texture_float_linear
WEBGL_compressed_texture_s3tc
WEBGL_compressed_texture_s3tc_srgb
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_lose_context
Audio information
System sample rate: 48000 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

Ogg Opus (audio/ogg; codecs=opus)
WebM Opus (audio/webm; codecs=opus)
Ogg Vorbis (audio/ogg; codecs=vorbis)
WebM Vorbis (audio/webm; codecs=vorbis)
MPEG-4 AAC (audio/mp4; codecs=mp4a.40.5)
MP3 (audio/mpeg)
FLAC (audio/flac)
PCM WAV (audio/wav; codecs=1)
Supported encode formats:

WebM Opus (audio/webm; codecs=opus)
Video information
Supported decode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
Ogg Theora (video/ogg; codecs=theora)
H.264 (video/mp4; codecs=avc1.42E01E)
Supported encode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)

</details>

**Comment by AshleyScirra** [comment]
The thread you linked appears to be a different problem to the one you described. The Apple forum thread that audio does not play in the *background*, whereas you are reporting an issue with audio not playing in the *foreground*. Can you confirm which issue you are actually having?

**Comment by Wackytoaster** [comment]
It‘s not playing at all, neither in back- nor foreground. The testapp I made shows the exact same error whenever I tap and the app tries to play a sound, that‘s why I found that thread. (I have the debug logs in the more details section)

```
2019-11-13 18:51:21.647210+0100 New project[1471:271866] [assertion] Error acquiring assertion: <NSError: 0x281c01c20; domain: RBSAssertionErrorDomain; code: 2; reason: "Client is missing required entitlement"> {
userInfo = {
RBSAssertionAttribute = <RBSLegacyAttribute: 0x15fde6db0; requestedReason: MediaPlayback; reason: MediaPlayback; flags: PreventTaskSuspend | PreventTaskThrottleDown | WantsForegroundResourcePriority>;
}
}
```

In the thread the user griv also reported the same thing.

> Same error here trying to play local html audio in a cordova app. Plays in Simulator not on device. Even in foreground.

Though I cannot check the simulator as it‘s still bugged.

**Comment by AshleyScirra** [comment]
Leaving this assigned to @shortercode for iOS testing but looking at the debug log I'd speculate that the latest iOS is much more aggressive about saving battery. Since media playback is likely to prevent the CPU falling back to low-power idle states due to requiring constant work, perhaps apps now have to opt-in to using a high-power state.

**Comment by AshleyScirra** [comment]
Just a heads up that iOS 13.2.3 is out now and might affect this if it's an iOS bug.

**Comment by shortercode** [comment]
Yeah... that's the version our mac doesn't want to debug at the moment helpfully. Just about to try replicating this on a 13.1.X device.

**Comment by shortercode** [comment]
So while I do get those messages about MedaPlayback in the log the device is still playing audio for me. They are definitely linked with audio playback as they are only emitted when the audio is started. I do not however get the ` [Process] kill() returned unexpected error 1` messages so that might be an indication that you are experiencing something different. Tied in with what @AshleyScirra said about the linked threads being about background audio only I'm not sure what else to investigate here.

I did additionally try adding the Background Audio and Background Processing capabilities to the project and it had no effect on the messages.

I think C2runtime projects will run on the simulator, as we don't require WASM for those. So you should be able to test that on the simulator, if you want to try validate it on something else.

Best hope is to probably try validate it on a 13.2.X device and see if they have resolved the problem. If I can get the mac talking to the 13.2.X device tomorrow I will update this.

**Comment by AshleyScirra** [comment]
iOS 13.2 came out ~3 weeks ago (Oct 28th) and according to StatCounter hasn't propagated past 15% yet. It's possible the original issue was a regression in iOS 13.2 which would explain why it doesn't reproduce on iOS 13.1.x, and potentially why Apple have held back on distribution of iOS 13.2 if there are known issues (and their release schedule has gone in to meltdown lately). It'd also explain why it was only just reported (iOS 13 originally came out mid-September and this is the first we've seen of this issue), and why we haven't had more reports (if all versions of iOS 13+ could not play audio, I'm sure we would have heard much more about it, but so far there's only this report). So yeah, I think the key thing is to test iOS 13.2.x, and verify if the latest update fixes it too - @Wackytoaster could you check with iOS 13.2.3?

**Comment by Wackytoaster** [comment]
@shortercode Funny enough, after updating from 13.2.2 to 13.2.3 it would also not allow me to connect to the device anymore (complaining about the device beeing locked when it defenitely wasn´t) but I fixed it by going Window > Devices and waiting there until the yellow bar gone away. Maybe it helps.

Here´s some info on the kill() error, that also seems to be something that started in 13.2.X
https://stackoverflow.com/questions/58321114/why-i-get-the-console-warning-process-kill-returned-unexpected-error-1-when

I´m using xCode 12.2.1 and tested the following:
iOS 12.4.1 (friends device, couldn´t debug though): Sound working
ipadOS 13.2.2: No sound
ipadOS 13.2.3: No sound
ipadOS 13.2.3 (Now with an entirely fresh build) still no sound

I never had a device with anything less than 13.2.2 as the ipad is brand new. Based on the thread I just assumed that it was anything 13+ but it seems that this isn´t actually the case. I´m not sure how far iOS and ipadOS differ from each other, maybe that could also be an issue.

Here´s one more report from 4 days ago mentioning 13.2.2
https://stackoverflow.com/questions/58887304/ios-13-webaudio-is-totally-broken-for-html-audio-elements-within-a-webview

**Comment by AshleyScirra** [comment]
Looks like a bug has been filed with Apple for this: https://bugs.webkit.org/show_bug.cgi?id=204261

It sounds pretty bad, looks like iOS 13.2.3 is still affected and so far there is no known workaround. We may have to wait for Apple to fix it... which I predict will be painful if they distribute an affected release to all devices, and the latest version is still affected... 😕

**Comment by Wackytoaster** [comment]
Bummer. Well I´ve included a note in the app on iOS next to the volume slider that sound **may** not work on devices with iOS 13+ but I guess in this case there isn´t much else to do other than wait. That´s still quite a terrible bug though.

**Comment by shortercode** [comment]
Thanks for the detailed feedback @Wackytoaster confirming the issue occurs on multiple devices and finding those issues was very helpful.

But mmm yeah sounds like Apple broke something big again. There was a mention on the issue that BufferSourceNodes are not affected. On iOS sounds are already being decoded to AudioBuffers, so if required could we use that @AshleyScirra? You know more about the WebAudio API than I do so I'm not sure what the implications are.

At least it sounds like they are aware of the issue and investigating it. But they don't have the best track record for actually getting fixes to users promptly.

**Comment by shortercode** [comment]
Marking as external for now, will remove the can't reproduce when I get the project running on a 13.2.X device.

**Comment by Wackytoaster** [comment]
> There was a mention on the issue that BufferSourceNodes are not affected
I´ve seen that but it was also mentioned that this won´t work for longer sounds. So that´s to keep in mind.

**Comment by AshleyScirra** [comment]
Let's get a device on 13.2.3 and test ourselves there first, then we can figure out what could be done.

**Comment by AshleyScirra** [comment]
I'd also add that Construct games in WKWebView have always used Web Audio only for audio playback. We could never get the `<audio>` element working due to the tight restrictions of WKWebView when loading local content. So despite the fact the Apple bug mentions the `<audio>` tag only, if it reproduces with Construct games, that indicates it's affecting all audio playback including Web Audio, since we never use `<audio>` anyway.

**Comment by shortercode** [comment]
Well it would appear that the iPhone I was going to test on is on a iOS 13.3 beta. Which does work, so that's good news I guess. There's a new beta version so I'm going to check that, then roll the device back to 13.2 just so we can validate the bug occurs on this device.

**Comment by shortercode** [comment]
I've tested that iPhone 7 on 13.2.3 and it still produces sound, @Wackytoaster did you verify that it occurred on an iPhone running 13.2.x or just iPad?

**Comment by Wackytoaster** [comment]
I could only test it on ipad myself. I actually did install the 13.3 beta too a few days ago but no change.

**Comment by shortercode** [comment]
Ok thanks for confirming. I'm updating an iPad to 13.2, will report back when I've checked.

**Comment by shortercode** [comment]
Okay so 13.2 also appears fine on the iPad. However, I have noted that toggling silent mode in the quick settings will block the audio. Even when the media volume is turned up.

**Comment by Wackytoaster** [comment]
The good news is that this is exactly the issue. Silent mode was on and switching it off did the trick. But wtf though... as you said this switch does **not** prevent webpages or other apps (youtube) from playing audio so it probably is a wkwebview thing? As such it also never crossed my mind that this could‘ve been the issue because sound generally worked.  What a bs issue :( I mean it’s still kind of a bug at least... 

A quick search for that issue even has some workarounds pop up.
https://stackoverflow.com/questions/35289918/play-audio-when-device-in-silent-mode-ios-swift/54253201

I hope you didn‘t sink too much time into it.

**Comment by shortercode** [comment]
Well thanks for confirming, it's good that we got a reproduction down at least. I decided to check against an iOS 12 device as well and can confirm the same issue exists there as well. We have had some mystery reports of iOS audio issues in the past so perhaps this is the explanation. Although I would have thought it would have been a more common issue.

I've been doing some additional research and the restriction appears to be down to 2 audio playback systems in iOS: AVAudioSessionCategoryAmbient and AVAudioSessionCategoryPlayback. The prior 
 implies to the OS that the Audio is optional, and hence silent mode will block is. The latter is not optional.

I found [this stackoverflow post](https://stackoverflow.com/questions/21122418/ios-webaudio-only-works-on-headphones/46839941#46839941) which suggests that WKWebView actually uses both forms of audio playback, for different parts. Primarily WebAudio is ambient and the HTMLAudioElement isn't. But it sounds like it might be possible to get WebAudio to work as well under special circumstances...

@AshleyScirra has a better knowledge of our audio system than me, what are your thoughts?
